File: kotlin-desktop-toolkit-win32/src/main/kotlin/org/jetbrains/desktop/win32/Application.kt
package org.jetbrains.desktop.win32

import org.jetbrains.desktop.win32.generated.NativeApplicationCallbacks
import org.jetbrains.desktop.win32.generated.NativeEventHandler
import org.jetbrains.desktop.win32.generated.desktop_windows_h
import java.lang.foreign.Arena
import java.lang.foreign.MemorySegment

public object Application {
    private var appPtr: MemorySegment? = null
    private var eventHandler: EventHandler? = null

    public fun init() {
        ffiDownCall {
            appPtr = Arena.ofConfined().use { arena ->
                desktop_windows_h.application_init(applicationCallbacks())
            }
        }
    }

    public fun runEventLoop(eventHandler: EventHandler) {
        ffiDownCall {
            this.eventHandler = eventHandler
            desktop_windows_h.application_run_event_loop(appPtr!!)
        }
    }

    public fun stopEventLoop() {
        ffiDownCall {
            desktop_windows_h.application_stop_event_loop(appPtr!!)
        }
    }

    public fun createWindow(params: WindowParams): Window {
        return Window.create(appPtr!!, params)
    }

    private fun runEventHandler(windowId: WindowId, event: Event): EventHandlerResult {
        return eventHandler?.let { eventHandler ->
            eventHandler(windowId, event)
        } ?: run {
            Logger.warn { "eventHandler is null; event: $event was ignored!" }
            EventHandlerResult.Continue
        }
    }

    // called from native
    private fun onEvent(windowId: WindowId, nativeEvent: MemorySegment): Boolean {
        return ffiUpCall(defaultResult = false) {
            val event = Event.fromNative(nativeEvent)
            val result = runEventHandler(windowId, event)
            when (result) {
                EventHandlerResult.Continue -> false
                EventHandlerResult.Stop -> true
            }
        }
    }

    private fun applicationCallbacks(): MemorySegment {
        val arena = Arena.global()
        val callbacks = NativeApplicationCallbacks.allocate(arena)
        NativeApplicationCallbacks.event_handler(callbacks, NativeEventHandler.allocate(::onEvent, arena))
        return callbacks
    }
}

File: native/desktop-windows/src/win32/application.rs
use std::rc::Rc;

use windows::core::Result as WinResult;

use super::{event_loop::EventLoop, events::EventHandler, window::Window, window_api::WindowParams};

pub struct Application {
    event_loop: Rc<EventLoop>,
}

impl Application {
    pub fn new(event_handler: EventHandler) -> WinResult<Self> {
        let event_loop = EventLoop::new(event_handler)?;
        Ok(Self {
            event_loop: Rc::new(event_loop),
        })
    }

    pub fn run_event_loop(&self) {
        self.event_loop.run();
    }

    pub fn shutdown(&self) -> WinResult<()> {
        self.event_loop.shutdown()
    }

    pub(crate) fn create_window(&self, params: &WindowParams) -> WinResult<Rc<Window>> {
        Window::new(params, Rc::downgrade(&self.event_loop))
    }
}
